### DISCUSSION TITLE
Whether to make the ligatures more fine-grained

### DISCUSSION THREAD

**Discussion by subframe7536** [subframe7536/maple-font#244] [Polls]
I'm contemplating whether it's worth modifying this font to a more generic one, see in #240

First of all, [Currently](https://github.com/subframe7536/maple-font/tree/v7.0-beta26?tab=readme-ov-file#features), the font is opinioned by default, like `a`, `@` or italic `l`, and use `cvXX` or `ssXX` to **disable and make the characters appear "correctly"**. Many people complains about the default style, and I have to reply and tell them how to enable the font features.

Also, the character style is inconsistent between regular and italic
![Clip_2024-09-08_09-25-09](https://github.com/user-attachments/assets/5c8f1897-668f-4262-83df-6b4daafb490c)

What's more, the current ligatures have already break the ligatures in V6, there is no more effect for users who migrate from V6

So I've initially developed the following modification plan which will introduce "breaking changes":

- Default characters: make them generic, just like `Jetbrains Mono`
- `cv01` - `cv04`: same as current, but the character style will become consistent between regular and italic
- `cv05`: italic alternative `l` `i` `j` `k` `x` (text is separated between `al`, `il`...)
- `ss01`: enable `==` `===` `!=` `!==` `=/=` ligatures
- `ss02`: enable arbitrary text tags, like `[todo]`
- `ss03`: enable `__` `#__` ligatures
- `ss04`: enable `>=` `<=` ligatures
- `ss05`: enable `\n` `\t`... ligatures (symbols like `\\` `\'` will be enabled in `calt`)
- `ss06`: enable `<<` `<<<` `>>` `>>>` ligatures
- `ss07`: enable italic `al` `el` `il` `ul` serials ligatures
- `calt`: others

If the changes are made, the release assets will contains two types of fonts:
- Base font: just like `Jetbrains Mono`, you can enable the features by yourself
- Default font: Custom build version, but everything is same as before, except font features

So there is no effect if you don't use font features, and more convenient for custom build

BTW, `ss01` - `ss04` are not changed, compare to V6

**Comment by matthew-e-brown** [comment]
I am not sure I would say the current font is "too" opinionated, but I would definitely appreciate having more fine-grained control over which ligatures appear when.

It would be convenient especially because I could configure my code editor to change to different ligatures depending on the language I'm writing in.

**Comment by arjpar** [comment]
I think it would be valuable to provide two assets for release, a default one where you can manually enable the features you want, and an opinionated one. If you have any questions @subframe7536 , feel free to ask ! Having more choices for users I think would benefit them and allows them to determine what they want.

**Comment by subframe7536** [comment]
What is blocking the release:

- Consistency: It appears that consistency is not as crucial as my initially thought.

- IDE/Terminal Support: There is limited support for configuring font features separately between regular and italic styles in IDEs or terminals.

- Naming: The naming convention for the prebuilt fonts is unclear. Should the opinionated version be the default? IMO, there is 3 options:
  - Keep same
  - `Maple Mono` (classic) and `Maple Mono Pro` (opinionated)
  - `Maple Mono` (opinionated) `Maple Mono Basic` (classic)?

**Reply by matthew-e-brown** [reply]
Since you created/designed the font, I think it makes sense for your opinionated one to be the default. New users see what you intended, but can customize it further if they'd like. Why not have the customizable one simply be called `Maple Mono Custom`? Using `Pro` or `Basic` makes it sound like one is _better_ than the other, but that's not really the case.

**Reply by subframe7536** [reply]
So now my thought is opposite to the subject: just extend current ligatures.

Disable the features by cvxx or ssxx is not a big problem.

Two versions of distribution maybe sometimes confusing other package maintainers.

Consistency has little help for users to memorize the font features.

**Reply by matthew-e-brown** [reply]
I think that as long as there is a way to custom-build a version with the ssxx and cvxx hard-coded (since not every editor/environment/document has support for advanced font-feature config), that would be totally fine.

I see that there is a `config.json` in the repo to control the build process, but is there any documentation on what to change `ignore` to in order to freeze the font features? 🤔

**Reply by subframe7536** [reply]
I plan to make a website just like [monaspace](https://monaspace.githubnext.com/) and there will have a visual config generator.

Also, if possible, Github Action is more convenient to custom build

**Comment by pkazmier** [comment]
I know I'm late to the party here, but I saw this discussion pinned, so I'll echo the comments of others. The default for **your** font should be **your** opinionated choices. As a user, I want to see how the font designer intended their font to be used. You've chosen the glyphs for a specific reason and I appreciate that.

With that said, I do think it's a great idea to provide the less opinionated version for folks that don't want or cannot set their own font features. For me, I love being able to tune, but I generally like to stick with what the designer intended. In the case of the current beta of V7, the only feature I've enabled is cv35 (non-cursive 'l').

**Comment by okkdev** [comment]
I'm very satisfied with the current features. The only thing I dislike is that arrows are included in cv01. I love the look of `@` with the gap, but dislike how arrows look with the little hole. It's nothing dramatic, I just keep cv01 enabled. Love the font, keep up the good work.

**Reply by subframe7536** [reply]
In fact, you can custom build it with `--apply-fea-file` and comment or delete these lines:

https://github.com/subframe7536/maple-font/blob/d59773ed0236e553ec1409c1b9f1ee04c884a7f7/source/features/regular.fea#L1620

https://github.com/subframe7536/maple-font/blob/d59773ed0236e553ec1409c1b9f1ee04c884a7f7/source/features/italic.fea#L1718

**Reply by okkdev** [reply]
Awesome! Thanks. I'll try that :)
